### ISSUE TITLE
nss-myhostname doesn't resolve the FQDN

### ISSUE THREAD

**Issue by primeos** [systemd/systemd#15894]
**Is your feature request related to a problem? Please describe.**

Yes, `hostname --fqdn` outputs only the hostname without the domain name (i.e. not the FQDN).

This is obviously not the intended behaviour. While this is fixable via entries in `/etc/hosts` doing this in `nss-myhostname` would be the best solution as this allows overriding that behaviour via `/etc/hosts` or the DNS (since `myhostname` should appear last in `/etc/nsswitch.conf`).

**Describe the solution you'd like**

`nss-myhostname` generates entries equivalent to the following entries in `/etc/hosts` (currently the `$FQDN` part is missing):
```
# Note: $FQDN comes first as it's the canonical hostname
127.0.0.2 $FQDN $HOSTNAME
::1 $FQDN $HOSTNAME
```

**Edit (forgot to mention):** `$FQDN` would be `kernel.hostname` appended by `kernel.domainname` (if set) while `$HOSTNAME` would be only `kernel.hostname`.

`getent hosts 127.0.0.2` shows that this is currently not the case. The relevant code is in `src/nss-myhostname/nss-myhostname.c`. Currently the hostname comes form [src/basic/hostname-util.c#L43](https://github.com/systemd/systemd/blob/v245/src/basic/hostname-util.c#L43) which corresponds to `uname --nodename` (which should come from the kernel parameter `kernel.hostname`). While it's possible to set `kernel.hostname` to the desired FQDN this seems not to be intended: [admin-guide/sysctl/kernel.html#domainname-hostname](https://www.kernel.org/doc/html/latest/admin-guide/sysctl/kernel.html#domainname-hostname).

**Describe alternatives you've considered**

`nss-resolve` (systemd-resolved) might already do the correct thing here (didn't test that yet), but not every distribution already uses systemd-resolved so `nss-myhostname` should also work as expected.

**Comment by poettering** [comment]
I am sorry, what?

FQDN is a DNS concept. The kernel only knows a single, generic "hostname" concept, not tied to DNS and thus typically that's a short one, without domains. Using one with domains ist not really such a great idea, since the definitions of what a valid DNS name is, and what the kernel allows are distinct (specifically, kernel hostnames can only be 64chars long, and DNS names 256).

kernel.domainname is generally understood to be a NIS thing, and not a DNS thing, and the kernel doesn't really define what it means, and doesn't really care much.

I don't really grok what you are asking for, but nss-myhostname is extremely simple, it just maps whatever you configured as kernel hostname string to the local IP addresses, that's all. If you use the fqdn as kernel hostname then it will map the fdn to the IP addresses, and if you use a short one, then it maps that. It's really just a string.

Hence, I really don't get what you are trying to do any what you want precisely?

**Comment by primeos** [comment]
tl;dr: I did some additional research and it seems like `nss-myhostname` cannot really do anything about this. I still think that it would be nice to have `hostname --fqdn` working as expected but we seem to have to rely on a proper (m)DNS (+ DHCP) setup for this (which seems to be what @poettering already briefly mentioned and what the confusion was about).

@poettering: Feel free to close this issue then. I replied to your questions below and added some additional notes. Feel free to have a quick look if you have time, but in the end I came to the conclusions above and as a result `systemd` most likely cannot really help much here. But I guess at least this is "documented" here now (I didn't find any existing issues regarding this).

---

> FQDN is a DNS concept.

Yes, that's unfortunately where it get's a bit problematic as current documentation does sometimes make a distinction here and sometimes not (IIRC). I only intended to use FQDN as a short form for `hostname+"."+domainname`. The main thing to discuss here is if we can make an assumption like domainname=dnsdomainname=ypdomainname=nisdomainname (on a practical level, as this is technically obviously not true).

Also, to avoid unnecessary confusion, I want to quickly state that I don't think the current `nss-myhostname` implementation is wrong (and therefore used the feature request template). However, IMO it would be nice to resolve the "FQDN" as well which is why I brought this up and would like to discuss this a bit.

> The kernel only knows a single, generic "hostname" concept, not tied to DNS and thus typically that's a short one, without domains. Using one with domains ist not really such a great idea, since the definitions of what a valid DNS name is, and what the kernel allows are distinct (specifically, kernel hostnames can only be 64chars long, and DNS names 256).

Agreed, `kernel.hostname` shouldn't contain the "domain part".

> kernel.domainname is generally understood to be a NIS thing, and not a DNS thing, and the kernel doesn't really define what it means, and doesn't really care much.

This is the thing I'd like to discuss here. Can we assume that `FQDN = kernel.hostname+kernel.domainname`? The era of NIS/YP was long before my time so I'm unfortunately not familiar with it (or what currently remains of it). But the more I look into this I'm starting to realize that this seems like a bad idea an that `kernel.domainname` is a bad fit for providing a domain name that we can also use as DNS domain name.

> Hence, I really don't get what you are trying to do any what you want precisely?

My goal was to enable applications to automatically determine the FQDN of the system (apparently e.g. [some MUAs/MTAs depend on this](https://github.com/NixOS/nixpkgs/issues/87633) without any application specific configuration (i.e. like `hostname --fqdn` does it). As multiple FQDNs can obviously be mapped to a system (which is another problem, especially in mobile environments)  I use FQDN here to refer to what `hosts(5)` calls the `canonical_hostname`. Using `/etc/hosts` an appropriate configuration could look like this (though some applications apparently assume that `127.0.0.1` resolves back to `localhost` - which is probably a good idea in general):
```
127.0.0.1 test.example.org test.example.com test localhost
::1 test.example.org test.example.com test localhost
```

A more appropriate `/etc/hosts` configuration could look like [this](https://gist.github.com/primeos/8f7f8e1e95518076ef38924125a2f921#file-hostname-test-c-L12).

However, after my additional research I think that:
1) `systemd` doesn't have a way to determine the (DNS) domain name.
    - Except if we'd use `kernel.domainname` which is technically not appropriate and most likely not used very widely (though I haven't looked into this yet).
    - This is the main problem that I missed before opening this issue (on NixOS we have the configuration option `networking.domain` for this, but there isn't something generic).
2) My idea to get the FQDN via `nss-myhostname` seems flawed and we should instead rely on a proper DNS (and DHCP) configuration to get the domain name part of the FQDN (i.e. a correct `domain` entry in `/etc/resolv.conf` and a properly configured local DNS resolver / mDNS setup).

**Comment by poettering** [comment]
i'd really stay away from the kernel domainname. As mentioned it's a NIS thing, and NIS is legacy stuff I would not touch, and certainly not expose new support for in 2020. 

Note that nss-systemd is fallback only, DNS is checked first, and so is /etc/hosts. i.e. it's the last resort check if you have nothing else set for the local system. If you want to configure an fqdn and make sure it remains resolvable even without DNS up then add it to /etc/hosts, that's totally an OK thing to do. 

Hence, let's close this, as I don't think there's anything actionable here.
